common/src/main/java/org/ergoplatform/appkit/Eip4Token.java

```diff
@@ -0,0 +1,229 @@
+        this(id, amount,
+            new String(JavaHelpers$.MODULE$.collToByteArray(r4.getValue()), StandardCharsets.UTF_8),
+            new String(JavaHelpers$.MODULE$.collToByteArray(r5.getValue()), StandardCharsets.UTF_8),
+            Integer.parseInt(new String(JavaHelpers$.MODULE$.collToByteArray(r6.getValue()), StandardCharsets.UTF_8)),
+            r7, r8, r9);
+    }
+
+    public Eip4Token(@Nonnull String id, long amount, @Nonnull String name,
+                     @Nonnull String description, int decimals) {
+        this(id, amount, name, description, decimals, null, null, null);
+    }
+
+    public Eip4Token(@Nonnull String id, long amount, @Nonnull String name,
+                     @Nonnull String description, int decimals,
+                     @Nullable ErgoValue<?> r7, @Nullable ErgoValue<?> r8, @Nullable ErgoValue<?> r9) {
+        super(id, amount);
+        this.name = name;
+        this.description = description;
+        this.decimals = decimals;
+
+        // these can be nullable, but either all of them or R7 and R8 defined
+        if (!(r7 != null && r8 != null || r7 == null && r8 == null && r9 == null)) {
+            throw new IllegalArgumentException("Either define all of R7 to R9 or none of them");
+        }
+
+        this.r7 = r7;
+        this.r8 = r8;
+        this.r9 = r9;
+    }
+
+    /**
+     * @return EIP-4 token verbose name (R4)
+     */
+    @Nonnull
+    public String getTokenName() {
+        return name;
+    }
+
+    /**
+     * @return EIP-4 token description (R5)
+     */
+    @Nonnull
+    public String getTokenDescription() {
+        return description;
+    }
+
+    /**
+     * @return EIP-4 token decimals (R6)
+     */
+    public int getDecimals() {
+        return decimals;
+    }
+
+    /**
+     * @return token amount taking decimals into account
+     */
+    @Nonnull
+    public String getAmountFormatted() {
+        return BigDecimal.valueOf(getValue()).movePointLeft(decimals).toPlainString();
+    }
+
+    @Nonnull
+    public AssetType getAssetType() {
+        byte[] assetType = getR7ByteArrayOrNull();
+        if (r7 == null) {
+            return AssetType.NONE;
+        } else if (assetType == null) {
+            return AssetType.UNKNOWN;
+        } else {
+            for (AssetType type : AssetType.values()) {
+                byte[] r7EntryForType = type.getR7ByteArrayForType();
+
+                if (r7EntryForType != null && Arrays.equals(r7EntryForType, assetType))
+                    return type;
+            }
+
+            return AssetType.UNKNOWN;
+        }
+    }
+
+    @Nullable
+    private byte[] getR7ByteArrayOrNull() {
+        // r7 specifies the asset type and should be Coll<Byte>
+        if (r7 != null && r7.getValue() instanceof Coll) {
+            Coll<?> assetType = (Coll<?>) r7.getValue();
+
+            if (assetType.length() > 0 && assetType.apply(0) instanceof Byte) {
```

Comment:
There is better way of checking the type of Coll elements by calling this method https://github.com/ScorexFoundation/sigmastate-interpreter/blob/d64fd2783c57e70a8ed045a24d2b195aad8eb1f7/library-api/src/main/scala/special/collection/Colls.scala#L374  and then comparing with https://github.com/ScorexFoundation/sigmastate-interpreter/blob/c63e8be72a772c39f1efb828ad1ee89a8deb5a9b/common/src/main/scala/scalan/TypeDesc.scala#L80

Or better yet, wrapping RType into ErgoType and then comparing with ErgoType.byteType()

Reply:
I have difficulties translating this to Java. Is this code already used from Java methods?

Reply:
Something like this 
```
    static public <T> ErgoValue<Coll<T>> of(T[] arr, ErgoType<T> tT) {
        Coll<T> value = JavaHelpers.SigmaDsl().Colls().fromArray(arr, tT.getRType());
        if (value.tItem() == ErgoType.byteType().getRType()) {
        ...
        }
        return new ErgoValue<>(value, ErgoType.collType(tT));
    }
```